### PR TITLE
Azure pipelines

### PR DESCRIPTION
**Pull request by nodeJayS**
Using Docker build to run dotnet test and sonar analysis

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      displayName: docker build image
+      inputs:
+        repository: 'posts'
+        command: 'build'
+        Dockerfile: '**/Dockerfile'
+    - task: DotNetCoreCLI@2
+      displayName: dotnet test
+      inputs:
+        command: 'test'
+        arguments: '--configuration $(BuildConfiguration) --collect:"XPlat Code Coverage" -- DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.Format=opencover'
```
Comment (escalonn) [inline azure-pipelines.yml:58]: `BuildConfiguration` is not `buildConfiguration`, i've corrected this error in like 4 other code reviews, pay more attention.
really we could just hardcode Release everywhere and drop the variable but it's fine

File: azure-pipelines.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        publishTestResults: true
+        workingDirectory: 'Fakebook.Posts/'
+    - task: SonarCloudAnalyze@1
+      displayName: sonar run analysis
```
Comment (escalonn) [inline azure-pipelines.yml:62]: need to build the whole solution (outside docker) before run analysis. dotnet test does not necessarily build the whole solution.

**Review by escalonn — CHANGES_REQUESTED** [review]
every one of these feedback comments i've already given to multiple other PRs. i shouldnt have to do that, people working on the CI and dockerfiles should be working together, sharing the feedback, and checking each other's approved code. it's very inefficient if they aren't

**Review by nodeJayS — COMMENTED** [review]
Revised pipelines accordingly

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        extraProperties: |
+          sonar.cs.opencover.reportsPaths=$(Agent.TempDirectory)/**/coverage.opencover.xml
+          sonar.cs.vstest.reportsPaths=$(Agent.TempDirectory)/*.trx
+    - task: UseDotNet@2
+      displayName: 'Use .NET 5 SDK'
+      inputs:
+        packageType: 'sdk'
+        version: '5.x'
+        installationPath: $(Agent.ToolsDirectory)/dotnet
+        workingDirectory: 'Fakebook.Posts.RestApi/'
```
Comment (escalonn) [inline azure-pipelines.yml:53]: this still doesn't belong
we dont care about the installationpath either for that matter

File: azure-pipelines.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    steps:
+    - task: Docker@2
+      displayName: Build Docker Image
+      inputs:
+        repository: $(imageName)
+        command: 'build'
+        Dockerfile: 'Fakebook.Posts/Dockerfile'
+# sonar cloud
+- stage: sonarCloud
```
Comment (escalonn) [inline azure-pipelines.yml:23]: this stage is not dependent on the previous one so use dependsOn like the other merged pipelines should be

File: azure-pipelines.yml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        version: 2.x
+    - task: SonarCloudPrepare@1
+      inputs:
+        SonarCloud: 'SonarCloud Token'
+        organization: '2011-fakebook-project3'
+        scannerMode: 'MSBuild'
+        projectKey: '2011-fakebook-project3_posts'
+        extraProperties: |
+          sonar.cs.opencover.reportsPaths=$(Agent.TempDirectory)/**/coverage.opencover.xml
+          sonar.cs.vstest.reportsPaths=$(Agent.TempDirectory)/*.trx
```
Comment (escalonn) [inline azure-pipelines.yml:46]: FYI we do want to get this trx -> sonarcloud in the other pipelines that don't have it but it's low priority.

File: azure-pipelines.yml (hunk 2 above)
Comment (escalonn) [reply azure-pipelines.yml:62]: this was not addressed

**Review by escalonn — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml [hunk 6]
```diff
@@ -43,23 +46,27 @@ stages:
         projectKey: '2011-fakebook-project3_posts'
         extraProperties: |
           sonar.cs.opencover.reportsPaths=$(Agent.TempDirectory)/**/coverage.opencover.xml
-          sonar.cs.vstest.reportsPaths=$(Agent.TempDirectory)/*.trx
```
Comment (escalonn) [inline azure-pipelines.yml]: no wait, sending the trx to sonarcloud is good, not bad. was it failing?

File: azure-pipelines.yml [hunk 7]
```diff
@@ -43,23 +46,27 @@ stages:
...
+
     - task: UseDotNet@2
       displayName: 'Use .NET 5 SDK'
       inputs:
-        packageType: 'sdk'
-        version: '5.x'
-        installationPath: $(Agent.ToolsDirectory)/dotnet
-        workingDirectory: 'Fakebook.Posts.RestApi/'
+        workingDirectory: '$(Agent.ToolsDirectory)/dotnet'
```
Comment (escalonn) [inline azure-pipelines.yml:53]: ??? workingDirectory is one input here that you don't need, and packageType and version the two that you _do_ need

File: azure-pipelines.yml [hunk 8]
```diff
@@ -43,23 +46,27 @@ stages:
...
+
+    - script: dotnet build --configuration $(buildConfiguration)
+      displayName: dotnet build
+      workingDirectory: $(targetSolution)
```
Comment (escalonn) [inline azure-pipelines.yml]: and yeah fix this like the error says

**Review by escalonn — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml (hunk 8 above)
Comment (escalonn) [reply azure-pipelines.yml]: $(directory) isn't right, the whole point of that step is to ensure we build the whole solution.
"directory" is a badvariable name for this reason, ambiguous

File: azure-pipelines.yml (hunk 6 above)
Comment (nodeJayS) [reply azure-pipelines.yml]: Re-added, assumed all pipelines would add it later together

File: azure-pipelines.yml (hunk 8 above)
Comment (nodeJayS) [reply azure-pipelines.yml]: removed directory variable

File: azure-pipelines.yml (hunk 3 above)
Comment (nodeJayS) [reply azure-pipelines.yml]: Removed installationPath and workingDirectory since they don't do anything

**Review by escalonn — CHANGES_REQUESTED** [review]
add `condition: always()` to the sonar run and sonar publish tasks, so that we get sonar analysis even when all the tests don't pass

**Review by escalonn — APPROVED** [review]

**Pull request merged**